File: sentimental-credit/credit.py
import cs50
import sys

# Getting input from user
num = cs50.get_int("Number: ")
length = len(str(num))
digits = []
i = 0
j = 1

# Listing all digits of number
while i < length:
    tmp = int(num / j) % 10
    digits.append(tmp)
    i += 1
    j *= 10

mul = 0

# Applying Luhn's Algorithm
# first step multiplying every second digit and adding them according to the Algorithm
for i in digits[1::2]:
    tmp = i * 2
    if len(str(tmp)) > 1:
        temp = tmp % 10
        mul += temp
        temp = int(tmp / 10) % 10
        mul += temp
    else:
        mul += tmp

add = mul

# Now adding all remaining digits from number
# If length is odd adding 1st digit
if length % 2 != 0:
    add += digits[-1]

# Else adding upto 2nd digit
for j in digits[:-1:2]:
    add += j

# Checking if Luhn's Algorithm is satisfied, if not returning INVALID
if add % 10 != 0:
    print("INVALID")
    sys.exit()

# Checking conditions for various credit cards
if ((length == 15) and (digits[-1] == 3)) and (digits[-2] == 4 or digits[-2] == 7):
    print("AMEX")
elif ((length == 16) and (digits[-1] == 5)) and (digits[-2] == 1 or digits[-2] == 2 or digits[-2] == 3 or digits[-2] == 4 or digits[-2] == 5):
    print("MASTERCARD")
elif (length == 16 or length == 13) and digits[-1] == 4:
    print("VISA")
else:
    print("INVALID")
    sys.exit()

File: credit/credit.c
#include <cs50.h>
#include <string.h>
#include <stdio.h>

int main(void)
{
    // Input card number
    long n = get_long("Number: ");
    // Calculate the length
    int i = 0;
    long l = n;
    while (l > 0)
    {
        l = l / 10;
        i++;
    }
    // Check whether length is valid
    if (i != 13 && i != 15 && i != 16)
    {
        printf("INVALID\n");
        return 0;
    }
    // Calculate sum
    int sum1 = 0;
    int sum2 = 0;
    long x = n;
    int total = 0;
    int d1;
    int d2;
    int mod1;
    int mod2;
    do
    {
        // Remove last digit
        mod1 = x % 10;
        x = x / 10;
        sum1 = sum1 + mod1;
        // Remove second last digit
        mod2 = x % 10;
        x = x / 10;
        // 2 times second last digit and add digits to sum2
        mod2 = mod2 * 2;
        d1 = mod2 % 10;
        d2 = mod2 / 10;
        sum2 = sum2 + d1 + d2;
    }
    while (x > 0);
    total = sum1 + sum2;
    // Checking Luhn Algorithm
    if (total % 10 != 0)
    {
        printf("INVALID\n");
        return 0;
    }
    // Starting digits
    long start = n;
    do
    {
        start = start / 10;
    }
    while (start > 100);
    // Checking starting digits for the card type
    if ((start / 10 == 5) && (0 < start % 10 && start % 10 < 6))
    {
        printf("MASTERCARD\n");
    }
    else if ((start / 10 == 3) && (start % 10 == 4 || start % 10 == 7))
    {
        printf("AMEX\n");
    }
    else if (start / 10 == 4)
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }
}